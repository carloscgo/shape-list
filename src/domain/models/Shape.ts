// domain/models/Shape.ts

import { Some } from '../../application'

export type IdShape = string | number

export type TypeShape = 'Square' | 'Circle' | 'Triangle'

export type Shape = {
  index: IdShape
  type: TypeShape
  [key: string]: Some
}
