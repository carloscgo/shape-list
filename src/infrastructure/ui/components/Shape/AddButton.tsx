import { memo } from 'react'
import { HiPlusCircle } from 'react-icons/hi2'
import { PropsAddButton } from '../../utils/interfaces'
import { SizeButton, SizeIcon } from '../../utils/constants'

const AddButton = memo(({ onAdd, ...props }: PropsAddButton) => {
  const { className, ...attrs } = props

  return (
    <div
      role="button"
      aria-label="add shape"
      className={`flex items-center justify-center cursor-pointer bg-white rounded-full w-[${SizeButton}] h-[${SizeButton}] mx-auto ${className}`}
      onClick={onAdd}
      {...attrs}
    >
      <HiPlusCircle size={SizeIcon} className="text-blue-600" />
    </div>
  )
})

export default AddButton
