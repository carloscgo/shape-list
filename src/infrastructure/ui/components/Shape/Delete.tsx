import { memo } from 'react'
import { HiOutlineTrash } from 'react-icons/hi2'
import { DragDropT, PropsDeleteDrop } from '../../utils/interfaces'

const Delete = memo(({ show, onDelete }: PropsDeleteDrop) => {
  const handleDrop = (e: DragDropT) => {
    e.preventDefault()
    e.stopPropagation()

    if (e.type === 'drop') {
      onDelete()
    }
  }

  if (!show) return null

  return (
    <div
      className="flex border-[1px] border-red-600 border-dashed rounded-full items-center justify-center w-[300px] h-[300px] mx-auto"
      onDragOver={handleDrop}
      onDrop={handleDrop}
    >
      <HiOutlineTrash size="150" className="text-red-600" />
    </div>
  )
})

export default Delete
