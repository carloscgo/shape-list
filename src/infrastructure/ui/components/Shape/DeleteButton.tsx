import { memo } from 'react'
import { HiOutlineTrash } from 'react-icons/hi2'
import { PropsDeleteButton } from '../../utils/interfaces'
import { SizeButton, SizeIcon } from '../../utils/constants'

const DeleteButton = memo(({ onDelete, ...props }: PropsDeleteButton) => {
  const { className, ...attrs } = props

  return (
    <div
      role="button"
      aria-label="delete shape"
      className={`flex items-center justify-center cursor-pointer bg-white rounded-full w-[${SizeButton}] h-[${SizeButton}] mx-auto ${className}`}
      onClick={onDelete}
      {...attrs}
    >
      <HiOutlineTrash size={SizeIcon} className="text-red-600" />
    </div>
  )
})

export default DeleteButton
