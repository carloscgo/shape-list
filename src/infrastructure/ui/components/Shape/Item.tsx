import { useEffect, useRef } from 'react'
import { useDrag, useDrop } from 'react-dnd'
import { ItemT, PropsShape } from '../../utils/interfaces'
import iconsShapes from '../../utils/iconShapes'

const type = 'Image'

const Item = ({ shape, index, moveShape, getIsMoved, onDrag }: PropsShape) => {
  const ref = useRef(null)

  const [, drop] = useDrop({
    accept: type,
    hover(item: ItemT) {
      if (!ref.current) {
        return
      }

      const dragIndex = item.index

      if (dragIndex === index) {
        return
      }

      moveShape(dragIndex, index)

      item.index = index
    },
  })

  const [{ isDragging }, drag] = useDrag(() => ({
    type: type,
    item: { id: shape.id, index },
    collect: (monitor) => {
      return {
        isDragging: monitor.isDragging(),
      }
    },
  }))

  drag(drop(ref))

  useEffect(() => {
    getIsMoved(isDragging)

    if (isDragging) {
      onDrag(shape)
    }
  }, [getIsMoved, isDragging, onDrag, shape])

  return (
    <div ref={ref} style={{ opacity: isDragging ? 0 : 1 }}>
      <div
        id={`${shape.type}-${shape.id}-${index}`}
        className="h-[200px] w-[200px] object-cover bg-transparent"
      >
        {iconsShapes[shape.type]('200px')}
      </div>
    </div>
  )
}

export default Item
