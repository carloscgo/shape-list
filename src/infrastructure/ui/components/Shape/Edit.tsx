// infrastructure/ui/components/Shape/New.tsx

import { useParams } from 'react-router-dom'
import { useEditShape, useGetShapeById } from '../../../../application'
import ShapesForm, { FormValues } from './Form'
import useToast from '../../hooks/useToast'
import { useTranslation } from '../../utils/i18n'
import { IdShape, Shape } from '../../../../domain/models/Shape'
import { useEffect, useState } from 'react'
import { Status } from '../../utils/constants'

const Edit = () => {
  const { idShape } = useParams()
  const [values, setValues] = useState<Shape>()
  const [status, setStatus] = useState<string | null>(null)
  const [error, setError] = useState<string | undefined>()

  const { t } = useTranslation()

  const editShapeAction = useEditShape()
  const getShapeByIdAction = useGetShapeById(idShape as IdShape)

  useEffect(() => {
    if (getShapeByIdAction) {
      getShapeByIdAction.then((data: Shape) => setValues(data))
    }
  }, [getShapeByIdAction])

  useToast(status, t('editingShape'), t('successfullyUpdated'), error)

  const onSubmit = (data: FormValues) => {
    setStatus(Status.loading)

    editShapeAction({ id: idShape, ...data })
      .then(() => setStatus(Status.success))
      .catch((error) => {
        setError(error)
        setStatus(Status.error)
      })
  }

  return (
    <ShapesForm
      title={t('editShape')}
      labelSubmit={t('editShape')}
      values={values}
      onSubmit={onSubmit}
    />
  )
}

export default Edit
