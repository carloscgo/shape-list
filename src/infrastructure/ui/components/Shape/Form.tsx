// infrastructure/ui/components/Shape/Form.tsx

import { type FieldValues, useForm } from 'react-hook-form'
import * as Yup from 'yup'
import { yupResolver } from '@hookform/resolvers/yup'

import Input from '../Input'
import FootButtons from '../FootButtons'
import { useTranslation } from '../../utils/i18n'
import routes from '../../utils/routes'
import { useEffect } from 'react'
import { Shape } from '../../../../domain/models/Shape'

export interface FormValues extends FieldValues {
  title: string
  reference: string
  description: string
  price: number
  tax: number
}

type PropsShapesForm = {
  title: string
  labelSubmit: string
  values?: Shape
  onSubmit: (data: FormValues, reset: () => void) => void
}

const Form = ({ title, labelSubmit, values, onSubmit }: PropsShapesForm) => {
  const { t } = useTranslation()

  const schema = Yup.object().shape({
    name: Yup.string().label(t('name')).trim().required().min(3).max(30),
    reference: Yup.string()
      .label(t('reference'))
      .trim()
      .required()
      .min(3)
      .max(30),
    description: Yup.string()
      .label(t('description'))
      .trim()
      .required()
      .max(100),
    price: Yup.number().label(t('price')).required(),
    tax: Yup.number().label(t('tax')).required(),
  })

  const {
    register,
    handleSubmit,
    formState: { errors, isValid, isSubmitting },
    setValue,
    reset,
  } = useForm<FormValues>({
    mode: 'onChange',
    resolver: yupResolver(schema),
  })

  useEffect(() => {
    if (typeof values === 'object' && values) {
      setValue('name', values.name)
      setValue('reference', values.reference)
      setValue('description', values.description)
      setValue('price', values.price as number)
      setValue('tax', values.tax as number)
    }
  }, [values, setValue])

  const callOnSubmit = (data: FormValues) => onSubmit(data, reset)

  return (
    <div className="container mx-auto max-w-md py-12">
      <h1 className="text-3xl font-medium my-5 dark:text-blue-300">{title}</h1>

      <form
        className="grid grid-cols-1 gap-y-6 shadow-lg p-8 rounded-lg"
        onSubmit={handleSubmit(callOnSubmit)}
      >
        <Input
          name="name"
          title={t('name')}
          register={register}
          errors={errors}
          validations={{ required: true, maxLength: 30 }}
        />
        <Input
          name="reference"
          title={t('reference')}
          register={register}
          errors={errors}
          validations={{ required: true, maxLength: 30 }}
        />
        <Input
          name="description"
          title={t('description')}
          register={register}
          errors={errors}
          validations={{ required: true, maxLength: 100 }}
        />
        <Input
          name="price"
          title={t('price')}
          type="number"
          step="0.01"
          min="0.1"
          register={register}
          errors={errors}
          validations={{ required: true, min: '0.1' }}
        />
        <Input
          name="tax"
          title={t('tax')}
          type="number"
          step="0.01"
          min="0.1"
          register={register}
          errors={errors}
          validations={{ required: true, min: '0.1' }}
        />

        <FootButtons
          routeBack={routes.list}
          isSubmitting={isSubmitting}
          isValid={isValid}
          label={labelSubmit}
        />
      </form>
    </div>
  )
}

export default Form
