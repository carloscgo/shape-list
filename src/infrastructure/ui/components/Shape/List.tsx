// infrastructure/ui/components/Shapes/List.tsx

import { memo, useEffect, useState } from 'react'
import orderBy from 'lodash/orderBy'

import { IdShape, Shape, TypeShape } from '../../../../domain/models/Shape'
import ShapeList from './ShapeList'
import { PositionsT } from '../../utils/interfaces'
import { useTranslation } from '../../utils/i18n'
import useShapes from '../../hooks/useShapes'

const List = memo(({ shapesInit = [] }: { shapesInit?: Shape[] }) => {
  const { t } = useTranslation()

  const [shapes, setShapes] = useState<Shape[]>(shapesInit)

  const { add, remove } = useShapes()

  const loadShapes = () =>
    setShapes((shapes) =>
      orderBy(shapes, ['index'], 'asc').map((shape: Shape) => ({
        index: shape.index,
        type: shape.type,
      }))
    )

  useEffect(() => {
    loadShapes()
  }, [])

  const actionDelete = (index: IdShape) => remove(index, shapes, setShapes)

  const actionAdd = (position: PositionsT, type: TypeShape, index: IdShape) =>
    add(index, position, type, shapes, setShapes)

  return (
    <div
      role="list"
      aria-label="container-main-shapes"
      className="w-full h-full flex flex-col items-center justify-center"
    >
      <div className="w-full h-10 mb-4 flex items-center justify-center bg-blue-600 dark:bg-blue-800 text-white">
        <h2 aria-label="header-title">{t('Figure Manager')}</h2>
      </div>

      <div className="bg-gray-100 dark:bg-[#1E293B] w-full h-full flex items-center justify-center shadow rounded-lg p-4 sm:p-6 xl:p-8 ">
        <ShapeList shapes={shapes} onAdd={actionAdd} onDelete={actionDelete} />
      </div>

      <div className="w-full h-5 mt-4 flex items-center justify-start text-black dark:text-white">
        <p>
          Developed by{' '}
          <a
            href="https://github.com/carloscgo/shape-list"
            className="hover:text-blue-700 underline"
            target="_blank"
            aria-label="developer"
          >
            Carlos Camacho
          </a>
        </p>
      </div>
    </div>
  )
})

export default List
