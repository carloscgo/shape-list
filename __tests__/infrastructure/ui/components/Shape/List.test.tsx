/* eslint-disable @typescript-eslint/no-unused-vars */
import {
  expect,
  describe,
  it,
  render,
  screen,
  fireEvent,
  beforeAll,
  React,
} from '../../../../../setupTest'
import { vi } from 'vitest'

import List from '../../../../../src/infrastructure/ui/components/Shape/List'
import { Shape } from '../../../../../src/domain/models/Shape'

describe('List', () => {
  const shapesInit = [{ index: 1, type: 'Circle' }] as Shape[]

  beforeAll(() => {
    global.ResizeObserver = class ResizeObserver {
      observe() {
        // do nothing
      }
      unobserve() {
        // do nothing
      }
      disconnect() {
        // do nothing
      }
    }
  })

  it('should render a list of shapes', () => {
    render(<List />)
    const shapeList = screen.getAllByRole('list')[0]
    expect(shapeList).toBeInTheDocument()
  })

  it('should call the onAdd callback when a shape is added', () => {
    vi.useFakeTimers()
    const { container, getAllByLabelText } = render(
      <List shapesInit={shapesInit} />
    )
    const item = getAllByLabelText('item')[0]
    fireEvent.mouseOver(item)
    vi.runAllTimers()
    const addBtn = getAllByLabelText('add shape')[0]
    fireEvent.click(addBtn)
    expect(container).toMatchSnapshot()
  })

  it('should call the onDelete callback when a shape is deleted', () => {
    vi.useFakeTimers()
    const { container, getAllByLabelText } = render(
      <List shapesInit={shapesInit} />
    )
    const item = getAllByLabelText('item')[0]
    fireEvent.mouseOver(item)
    vi.runAllTimers()
    const deleteBtn = getAllByLabelText('delete shape')[0]
    fireEvent.click(deleteBtn)
    expect(container).toMatchSnapshot()
  })
})
