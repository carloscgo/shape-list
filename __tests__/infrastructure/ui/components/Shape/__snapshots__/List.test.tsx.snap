// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List > should call the onAdd callback when a shape is added 1`] = `
<div
  aria-hidden="true"
>
  <div
    aria-label="container-main-shapes"
    class="w-full h-full flex flex-col items-center justify-center"
    role="list"
  >
    <div
      class="w-full h-10 mb-4 flex items-center justify-center bg-blue-600 dark:bg-blue-800 text-white"
    >
      <h2
        aria-label="header-title"
      >
        Figure Manager
      </h2>
    </div>
    <div
      class="bg-gray-100 dark:bg-[#1E293B] w-full h-full flex items-center justify-center shadow rounded-lg p-4 sm:p-6 xl:p-8 "
    >
      <div
        aria-label="container shapes"
        class="flex w-full p-10 overflow-auto"
        role="list"
      >
        <div
          aria-label="item-container"
          class="relative h-auto w-[300px] flex items-center justify-center px-2 left-0 top-0"
          data-test="item-container"
        >
          <div
            class="absolute left-[50%] right-[50%] flex items-center justify-center h-auto w-autp"
            style="transition: scale 2s; transform: translate(-50%, -50%); scale: 1; display: flex;"
          >
            <div
              aria-label="add shape"
              class="flex items-center justify-center cursor-pointer bg-white rounded-full w-[50px] h-[50px] mx-auto mt-40 mr-5 ml-0"
              data-test="add-button"
              role="button"
            >
              <svg
                aria-hidden="true"
                class="text-blue-600"
                fill="currentColor"
                height="40px"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="40px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M12 2.25c-5.385 0-9.75 4.365-9.75 9.75s4.365 9.75 9.75 9.75 9.75-4.365 9.75-9.75S17.385 2.25 12 2.25zM12.75 9a.75.75 0 00-1.5 0v2.25H9a.75.75 0 000 1.5h2.25V15a.75.75 0 001.5 0v-2.25H15a.75.75 0 000-1.5h-2.25V9z"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
            <div
              aria-label="delete shape"
              class="flex items-center justify-center cursor-pointer bg-white rounded-full w-[50px] h-[50px] mx-auto undefined"
              data-test="delete-button"
              role="button"
            >
              <svg
                aria-hidden="true"
                class="text-red-600"
                fill="none"
                height="40px"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                width="40px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14.74 9l-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 01-2.244 2.077H8.084a2.25 2.25 0 01-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 00-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 013.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 00-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 00-7.5 0"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </div>
            <div
              aria-label="add shape"
              class="flex items-center justify-center cursor-pointer bg-white rounded-full w-[50px] h-[50px] mx-auto mt-40 ml-5 mr-0"
              data-test="add-button"
              role="button"
            >
              <svg
                aria-hidden="true"
                class="text-blue-600"
                fill="currentColor"
                height="40px"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="40px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M12 2.25c-5.385 0-9.75 4.365-9.75 9.75s4.365 9.75 9.75 9.75 9.75-4.365 9.75-9.75S17.385 2.25 12 2.25zM12.75 9a.75.75 0 00-1.5 0v2.25H9a.75.75 0 000 1.5h2.25V15a.75.75 0 001.5 0v-2.25H15a.75.75 0 000-1.5h-2.25V9z"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
          <div
            aria-label="item"
            class="h-[200px] w-[200px] object-cover bg-transparent"
          >
            <svg
              class="text-teal-700"
              fill="none"
              height="200px"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="200px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12 17C14.7614 17 17 14.7614 17 12C17 9.23858 14.7614 7 12 7C9.23858 7 7 9.23858 7 12C7 14.7614 9.23858 17 12 17ZM12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
        <div
          style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
        />
      </div>
    </div>
    <div
      class="w-full h-5 mt-4 flex items-center justify-start text-black dark:text-white"
    >
      <p>
        Developed by
         
        <a
          aria-label="developer"
          class="hover:text-blue-700 underline"
          href="https://github.com/carloscgo/shape-list"
          target="_blank"
        >
          Carlos Camacho
        </a>
      </p>
    </div>
  </div>
</div>
`;

exports[`List > should call the onDelete callback when a shape is deleted 1`] = `
<div>
  <div
    aria-label="container-main-shapes"
    class="w-full h-full flex flex-col items-center justify-center"
    role="list"
  >
    <div
      class="w-full h-10 mb-4 flex items-center justify-center bg-blue-600 dark:bg-blue-800 text-white"
    >
      <h2
        aria-label="header-title"
      >
        Figure Manager
      </h2>
    </div>
    <div
      class="bg-gray-100 dark:bg-[#1E293B] w-full h-full flex items-center justify-center shadow rounded-lg p-4 sm:p-6 xl:p-8 "
    >
      <div
        aria-label="container shapes"
        class="flex w-full p-10 overflow-auto"
        role="list"
      >
        <div
          aria-label="add shape"
          class="flex items-center justify-center cursor-pointer bg-white rounded-full w-[50px] h-[50px] mx-auto my-20 mx-auto"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="text-blue-600"
            fill="currentColor"
            height="40px"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="40px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12 2.25c-5.385 0-9.75 4.365-9.75 9.75s4.365 9.75 9.75 9.75 9.75-4.365 9.75-9.75S17.385 2.25 12 2.25zM12.75 9a.75.75 0 00-1.5 0v2.25H9a.75.75 0 000 1.5h2.25V15a.75.75 0 001.5 0v-2.25H15a.75.75 0 000-1.5h-2.25V9z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="w-full h-5 mt-4 flex items-center justify-start text-black dark:text-white"
    >
      <p>
        Developed by
         
        <a
          aria-label="developer"
          class="hover:text-blue-700 underline"
          href="https://github.com/carloscgo/shape-list"
          target="_blank"
        >
          Carlos Camacho
        </a>
      </p>
    </div>
  </div>
</div>
`;
